internal/c/parts/video/font/font.cpp

```diff
@@ -1,339 +1,1417 @@
+        // This will help us quickly allocate a free handle
+        for (h = lowestFreeHandle; h < vectorSize; h++) {
+            if (!fonts[h]->isUsed) {
+                FONT_DEBUG_PRINT("Recent font handle %i recycled", h);
+                break;
             }
-        } else {
-            if (ox < 0) { // regress codepoint_ox?
-                if ((codepoint_ox + ox) >= 0) {
-                    codepoint_ox += ox;
-                } else {
-                    codepoint_ox = 0;
+        }
+
+        if (h >= vectorSize) {
+            // Scan through the entire vector and return a slot that is not being used
+            // Ideally this should execute in extremely few (if at all) senarios
+            // Also, this loop should not execute if size is 0
+            for (h = 0; h < vectorSize; h++) {
+                if (!fonts[h]->isUsed) {
+                    FONT_DEBUG_PRINT("Font handle %i recycled", h);
+                    break;
                 }
             }
         }
-        render[codepoint_i].data = data2;
-        render[codepoint_i].w = w2;
-        render[codepoint_i].ox = codepoint_ox;
-        codepoint_ox += w2;
-        if (codepoint_ox > codepoint_w)
-            codepoint_w = codepoint_ox;
-
-    } // codepointi loop
-
-    // join&'blend' multiple codepoints
-    w2 = codepoint_w;
-    h2 = fonts[i].default_pixel_height;
-    data2 = (uint8_t *)calloc(w2 * h2, 1);
-    for (codepoint_i = 0; codepoint_i < codepoints; codepoint_i++) {
-
-        data1 = render[codepoint_i].data;
-        w1 = render[codepoint_i].w;
-        h1 = h2;
-
-        ox2 = render[codepoint_i].ox;
-
-        for (y1 = 0; y1 < h1; y1++) {
-            y2 = y1;
-            for (x1 = 0; x1 < w1; x1++) {
-                x2 = x1 + ox2;
-                value = data1[x1 + y1 * w1];
-                if (value > data2[x2 + y2 * w2])
-                    data2[x2 + y2 * w2] = value;
-            } // x1
-        }     // y1
-        free(data1);
-
-    } // codepoint_i
-
-    *out_data = data2;
-    *out_x = w2;
-    *out_y = h2;
-    // Note: '*out_x_pre_increment' is set above
-    *out_x_post_increment = 0;
-    if (codepoints > 1)
-        free(render);
-    return 1;
+
+        if (h >= vectorSize) {
+            // If we have reached here then either the vector is empty or there are no empty slots
+            // Simply create a new handle at the back of the vector
+            auto newHandle = new Font; // allocate and initialize
+
+            if (!newHandle)
+                return -1; // we cannot return 0 here since 0 is a valid internal handle
+
+            fonts.push_back(newHandle);
+            size_t newVectorSize = fonts.size();
+
+            // If newVectorSize == vectorSize then push_back() failed
+            if (newVectorSize <= vectorSize) {
+                delete newHandle;
+                return -1; // we cannot return 0 here since 0 is a valid internal handle
```

Comment:
Could you use `INVALID_FONT_HANDLE` for these cases?

Reply:
`INVALID_FONT_HANDLE` is defined as zero because that is what stuff in `libqb.cpp` expects. However, `CreateHandle()` returns 0 as a valid handle and that handle is reserved in the constructor and never passed to libqb.cpp.

Reply:
Ah, thanks, I didn't catch that it was zero :)